### ISSUE TITLE
Vsphere: Terraform apply not able to complete - hangs on creating VM

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-vsphere#104]
_This issue was originally opened by @Joe6547654 as hashicorp/terraform#15504. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

# PROVIDER: VMWare vSphere
Hi,
I am trying to create a new VM from the template using terraform. Since guest OS I use is in the guest OS support matrix of VMware, so I add the "skip_customization". However, when I do a "terraform apply". Terraform gets stuck on the creating the VM state and doesn't move forward.  

### Terraform Configuration Files
```hcl
# Configure the VMware vSphere Provider
provider "vsphere" {
  user           = "${var.vsphere_user}"
  password       = "${var.vsphere_password}"
  vsphere_server = "${var.vsphere_server}"
  allow_unverified_ssl = true
}

# Create a folder
  resource "vsphere_folder" "Terraform" {
  datacenter = "Datacenter"
  path = "DSM/Terraform"
}

# Create a virtual machine within the folder
  resource "vsphere_virtual_machine" "test" {
  name   = "Terraform-test"
  folder = "${vsphere_folder.Terraform.path}"
  vcpu   = 2
  memory = 4096
  datacenter = "Datacenter"

  network_interface {
      label = "network"
  }
  disk {
     # Guest OS: Ubuntu Linux (32-bit)
     template  = "DSM/A/template"  
  }
   skip_customization = true
}
```
### Terrafrom Apply
This is what I get after terraform apply: (it shouldn't take that long, and it won't stop if I don't terminate it)
>...
> vsphere_virtual_machine.test: Still creating... (10s elapsed)
> vsphere_virtual_machine.test: Still creating... (20s elapsed)
> vsphere_virtual_machine.test: Still creating... (30s elapsed)
> vsphere_virtual_machine.test: Still creating... (40s elapsed)
> vsphere_virtual_machine.test: Still creating... (50s elapsed)
> ...
>vsphere_virtual_machine.test: Still creating... (21m20s elapsed)
> vsphere_virtual_machine.test: Still creating... (21m30s elapsed)
> vsphere_virtual_machine.test: Still creating... (21m40s elapsed)
> vsphere_virtual_machine.test: Still creating... (21m50s elapsed)
> vsphere_virtual_machine.test: Still creating... (22m0s elapsed)
> Interrupt received.
> Please wait for Terraform to exit or data loss may occur.
> Gracefully shutting down...
> stopping apply operation...
> vsphere_virtual_machine.test: Still creating... (22m10s elapsed)
> Two interrupts received. Exiting immediately. Note that data loss may have occurred.

### Expected Behavior
Finish creating and show message apply complete.

### Actual Behavior
Terraform hangs on creating VM. 

### Steps to Reproduce
1. `terraform plan`
2. `terraform apply`

### Important Factoids
Using user account with datacenter level permission access

**Comment by laxin84** [comment]
I'll add that I'm also experiencing this issue using Terraform 0.9.8, with a vSphere v5.5 endpoint.  As a bit of additional information, the terraform apply does, in fact, successfully create and customize a VM in the vSphere environment using an existing template and datastore.  vSphere just never seems to report back to Terraform that the operation succeeded, so the tfstate file is never properly updated and the infrastructure remains unmanaged.  In order to clean up the VM, manual destruction is necessary.

**Comment by alexissavin** [comment]
Same issue here with "Terraform v0.9.11" and vSphere 6.0.0.

VM deployment ends up successfully in vcenter, but terraform remain unable to complete its own process.
Message is : vsphere_virtual_machine.tfvm: Still creating... (17m0s elapsed)

**Comment by alekssaul** [comment]
I've seen this when the template is cloned but the VM gets powered on with network adapter disconnected. Is this the case as well? If so does connecting the VM's vnic manually signals back to terraform?

**Comment by alexissavin** [comment]
VM Nic is correctly activated but I do see the following in the debug logs

2017/07/12 17:17:48 [DEBUG] plugin: terraform: vsphere-provider (internal) 2017/07/12 17:17:48 [DEBUG] VM customization finished
2017/07/12 17:17:49 [DEBUG] plugin: terraform: vsphere-provider (internal) 2017/07/12 17:17:49 [INFO] Created virtual machine: tfvm
2017/07/12 17:17:49 [DEBUG] plugin: terraform: vsphere-provider (internal) 2017/07/12 17:17:49 [DEBUG] virtual machine resource data: &schema.ResourceData{schema:map[string]*schema.Schema{"memory_reservation":(*schema.Schema)(0xc420807c20), "linked_clone":(*schema.Schema)(0xc420808000), "enable_disk_uuid":(*schema.Schema)(0xc420808690), "windows_opt_config":(*schema.Schema)(0xc420808a50), "folder":(*schema.Schema)(0xc420807950), "dns_servers":(*schema.Schema)(0xc4208084b0), "moid":(*schema.Schema)(0xc420808870), "custom_configuration_parameters":(*schema.Schema)(0xc420808960), "domain":(*schema.Schema)(0xc4208081e0), "resource_pool":(*schema.Schema)(0xc420807ef0), "dns_suffixes":(*schema.Schema)(0xc4208083c0), "skip_customization":(*schema.Schema)(0xc4208085a0), "uuid":(*schema.Schema)(0xc420808780), "disk":(*schema.Schema)(0xc420808c30), "detach_unknown_disks_on_delete":(*schema.Schema)(0xc420808d20), "memory":(*schema.Schema)(0xc420807b30), "vcpu":(*schema.Schema)(0xc420807a40), "datacenter":(*schema.Schema)(0xc420807d10), "cluster":(*schema.Schema)(0xc420807e00), "gateway":(*schema.Schema)(0xc4208080f0), "time_zone":(*schema.Schema)(0xc4208082d0), "network_interface":(*schema.Schema)(0xc420808b40), "cdrom":(*schema.Schema)(0xc420808e10), "name":(*schema.Schema)(0xc420807860)}, config:(*terraform.ResourceConfig)(nil), state:(*terraform.InstanceState)(nil), diff:(*terraform.InstanceDiff)(0xc420892320), meta:map[string]interface {}(nil), timeouts:(*schema.ResourceTimeout)(0xc420838a20), multiReader:(*schema.MultiLevelFieldReader)(0xc420892740), setWriter:(*schema.MapFieldWriter)(0xc4208926a0), newState:(*terraform.InstanceState)(0xc4205ccd20), partial:false, partialMap:map[string]struct {}(nil), once:sync.Once{m:sync.Mutex{state:0, sema:0x0}, done:0x1}, isNew:true}
2017/07/12 17:17:50 [DEBUG] plugin: terraform: vsphere-provider (internal) 2017/07/12 17:17:50 [DEBUG] Set the moid: "vm-3687"
**2017/07/12 17:17:50 [DEBUG] plugin: terraform: vsphere-provider (internal) 2017/07/12 17:17:50 [DEBUG] Waiting for interfaces to appear**
2017/07/12 17:17:52 [DEBUG] dag/walk: vertex "meta.count-boundary (count boundary fixup)", waiting for: "vsphere_virtual_machine.tfvm"
2017/07/12 17:17:52 [DEBUG] dag/walk: vertex "provider.vsphere (close)", waiting for: "vsphere_virtual_machine.tfvm"
2017/07/12 17:17:52 [DEBUG] dag/walk: vertex "root", waiting for: "meta.count-boundary (count boundary fixup)"
2017/07/12 17:17:57 [DEBUG] dag/walk: vertex "provider.vsphere (close)", waiting for: "vsphere_virtual_machine.tfvm"
2017/07/12 17:17:57 [DEBUG] dag/walk: vertex "meta.count-boundary (count boundary fixup)", waiting for: "vsphere_virtual_machine.tfvm"
2017/07/12 17:17:57 [DEBUG] dag/walk: vertex "root", waiting for: "meta.count-boundary (count boundary fixup)"

**Comment by alexissavin** [comment]
Well, digging into this issue, it appears that the interface was connected on the vmware side, but disabled in the guest system.

The current behavior of the provider get stuck until the vSwitch port gets UP.

**Comment by alekssaul** [comment]
Np.. do you know if the ova has outdated vm tools or any modifications to vmtools systemd unit?

We've seen this in CoreOS linux where we were overriding the systemd unit  via cloud-init/ignition. Removing systemd overrides made the issue disappear.

**Comment by alexissavin** [comment]
Template was running on Centos 7.3 running open-vm-tools from official repository.

**Comment by playniuniu** [comment]
Can you see the IP address in the vshpere web client when vm is up.

Any way, you can chose 

1. Delete /etc/udev/rules.d/70-persistent-net.rules file
2. Disable IPv6 in CentOS

and check could it be able to fix the problem

**Comment by vancluever** [comment]
Hey all,

I think this issue is kind of old and possibly fixed now that v0.3.0 is out with the new customization and network waiter code.

If everyone in this thread that's having issues could give it a go, and put in a new issue if you are still seeing problems, that would be great. I'm going to close this one now as I think any related issues that will be arising will more than likely be newer issues.

Remember as well that [`wait_for_guest_net`](https://www.terraform.io/docs/providers/vsphere/r/virtual_machine.html#wait_for_guest_net) has now been added as well, so you can set this to `false` if there's any doubt that your VM will not be coming up with a valid default gateway or IP address.

Thanks!
